inotify watch limit reached

**Describe the bug**
Not sure what exactly generated the error, but it happened when I added new images in the datastore folder while the server was running. Now I can't start the server again

**To Reproduce**
Steps to reproduce the behavior:
1. Go to '...'
2. Install '....'
3. Run commands '....'

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**

**Environment**

Ensuring you use the relevant python executable, please paste the output of:

```
python -c 'import monai; monai.config.print_debug_info()'
```

**Additional context**

  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/starlette/routing.py", line 526, in lifespan
    async for item in self.lifespan_context(app):
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/starlette/routing.py", line 467, in default_lifespan
    await self.startup()
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/starlette/routing.py", line 502, in startup
    await handler()
  File "main.py", line 52, in startup_event
    get_app_instance()
  File "/home/adp20local/Documents/MONAILabel/monailabel/utils/others/generic.py", line 89, in get_app_instance
    return app_instance(settings.APP_DIR, settings.STUDIES)
  File "/home/adp20local/Documents/MONAILabel/monailabel/utils/others/app_utils.py", line 32, in app_instance
    o = c(app_dir=app_dir, studies=studies)
  File "/home/adp20local/Documents/MONAILabel/sample-apps/deepedit_heart/main.py", line 73, in __init__
    resources=None,
  File "/home/adp20local/Documents/MONAILabel/monailabel/interfaces/app.py", line 42, in __init__
    self._datastore: Datastore = LocalDatastore(studies)
  File "/home/adp20local/Documents/MONAILabel/monailabel/utils/datastore/local.py", line 110, in __init__
    self._observer.start()
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/observers/api.py", line 256, in start
    emitter.start()
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/utils/__init__.py", line 93, in start
    self.on_thread_start()
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/observers/inotify.py", line 118, in on_thread_start
    self._inotify = InotifyBuffer(path, self.watch.is_recursive)
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/observers/inotify_buffer.py", line 35, in __init__
    self._inotify = Inotify(path, recursive)
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/observers/inotify_c.py", line 199, in __init__
    self._add_dir_watch(path, recursive, event_mask)
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/observers/inotify_c.py", line 396, in _add_dir_watch
    self._add_watch(path, mask)
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/observers/inotify_c.py", line 417, in _add_watch
    Inotify._raise_error()
  File "/home/adp20local/anaconda3/envs/monai_env/lib/python3.6/site-packages/watchdog/observers/inotify_c.py", line 429, in _raise_error
    raise OSError(errno.ENOSPC, "inotify watch limit reached")
OSError: [Errno 28] inotify watch limit reached

[2021-05-21 16:27:20,183] [ERROR] (uvicorn.error) - Application startup failed. Exiting.

--- a/monailabel/interfaces/datastore.py
+++ b/monailabel/interfaces/datastore.py
@@ -1,6 +1,6 @@
 from abc import ABCMeta, abstractmethod
 from enum import Enum
-from typing import Any, Dict, List
+from typing import Any, Dict, List, Union
 
 
 class DefaultLabelTag(Enum):
@@ -56,7 +56,7 @@ class Datastore(metaclass=ABCMeta):
         pass
 
     @abstractmethod
-    def get_labels_by_image_id(self, image_id: str, tag: str = None) -> Dict[str, str]:
+    def get_labels_by_image_id(self, image_id: str, tag: str = None) -> Union[Dict[str, str], str]:
         """
         Retrieve all label ids for the given image id
 

--- a/monailabel/interfaces/datastore.py
+++ b/monailabel/interfaces/datastore.py
@@ -56,7 +56,7 @@ class Datastore(metaclass=ABCMeta):
         pass
 
     @abstractmethod
-    def get_labels_by_image_id(self, image_id: str) -> Dict[str, str]:
+    def get_labels_by_image_id(self, image_id: str, tag: str = None) -> Dict[str, str]:
         """
         Retrieve all label ids for the given image id
 

--- a/monailabel/utils/datastore/local.py
+++ b/monailabel/utils/datastore/local.py
@@ -113,7 +113,9 @@ class LocalDatastore(Datastore):
                 self._observer.schedule(self._handler, recursive=True, path=self._datastore_path)
                 self._observer.start()
             except OSError as e:
-                logger.error(f'File watcher limit reached in operating system. Local datastore will not update if images and labels are moved from datastore location.')
+                logger.error(
+                    "File watcher limit reached in operating system. Local datastore will not update if images and labels are moved from datastore location."
+                )
                 logger.error(str(e))
 
     def name(self) -> str:

--- a/monailabel/utils/datastore/local.py
+++ b/monailabel/utils/datastore/local.py
@@ -262,9 +262,9 @@ class LocalDatastore(Datastore):
         for obj in self._datastore.objects:
             if obj.image.id == image_id:
                 if tag is not None:
-                    labels = {label.id: label.tag for label in obj.labels}
-                else:
                     labels = {label.id: label.tag for label in obj.labels if label.tag == tag}
+                else:
+                    labels = {label.id: label.tag for label in obj.labels}
                 return labels
         return {}
 

--- a/monailabel/utils/datastore/local.py
+++ b/monailabel/utils/datastore/local.py
@@ -252,7 +252,7 @@ class LocalDatastore(Datastore):
                     return os.path.join(self._datastore_path, self._label_store_path, label.id)
         return ""
 
-    def get_labels_by_image_id(self, image_id: str) -> Dict[str, str]:
+    def get_labels_by_image_id(self, image_id: str, tag: str = None) -> Dict[str, str]:
         """
         Retrieve all label ids for the given image id
 
@@ -261,7 +261,10 @@ class LocalDatastore(Datastore):
         """
         for obj in self._datastore.objects:
             if obj.image.id == image_id:
-                labels = {label.id: label.tag for label in obj.labels}
+                if tag is not None:
+                    labels = {label.id: label.tag for label in obj.labels}
+                else:
+                    labels = {label.id: label.tag for label in obj.labels if label.tag == tag}
                 return labels
         return {}
 

--- a/monailabel/utils/datastore/local.py
+++ b/monailabel/utils/datastore/local.py
@@ -113,7 +113,7 @@ class LocalDatastore(Datastore):
                 self._observer.schedule(self._handler, recursive=True, path=self._datastore_path)
                 self._observer.start()
             except OSError as e:
-                logger.error(f'Watch limit reached in operating system. Local datastore will not update if images and labels are moved from datastore location.')
+                logger.error(f'File watcher limit reached in operating system. Local datastore will not update if images and labels are moved from datastore location.')
                 logger.error(str(e))
 
     def name(self) -> str:

--- a/monailabel/utils/datastore/local.py
+++ b/monailabel/utils/datastore/local.py
@@ -108,9 +108,13 @@ class LocalDatastore(Datastore):
                 ignore_patterns=[self._datastore_config_path],
             )
             self._handler.on_any_event = self._on_any_event
-            self._observer = Observer()
-            self._observer.schedule(self._handler, recursive=True, path=self._datastore_path)
-            self._observer.start()
+            try:
+                self._observer = Observer()
+                self._observer.schedule(self._handler, recursive=True, path=self._datastore_path)
+                self._observer.start()
+            except OSError as e:
+                logger.error(f'Watch limit reached in operating system. Local datastore will not update if images and labels are moved from datastore location.')
+                logger.error(str(e))
 
     def name(self) -> str:
         """

--- a/monailabel/utils/datastore/local.py
+++ b/monailabel/utils/datastore/local.py
@@ -6,7 +6,7 @@ import os
 import pathlib
 import shutil
 from pathlib import Path
-from typing import Any, Dict, List
+from typing import Any, Dict, List, Union
 
 from pydantic import BaseModel
 from watchdog.events import PatternMatchingEventHandler
@@ -114,7 +114,8 @@ class LocalDatastore(Datastore):
                 self._observer.start()
             except OSError as e:
                 logger.error(
-                    "File watcher limit reached in operating system. Local datastore will not update if images and labels are moved from datastore location."
+                    "File watcher limit reached in operating system. "
+                    "Local datastore will not update if images and labels are moved from datastore location."
                 )
                 logger.error(str(e))
 
@@ -252,7 +253,7 @@ class LocalDatastore(Datastore):
                     return os.path.join(self._datastore_path, self._label_store_path, label.id)
         return ""
 
-    def get_labels_by_image_id(self, image_id: str, tag: str = None) -> Dict[str, str]:
+    def get_labels_by_image_id(self, image_id: str, tag: str = None) -> Union[Dict[str, str], str]:
         """
         Retrieve all label ids for the given image id
 
@@ -262,10 +263,12 @@ class LocalDatastore(Datastore):
         for obj in self._datastore.objects:
             if obj.image.id == image_id:
                 if tag is not None:
-                    labels = {label.id: label.tag for label in obj.labels if label.tag == tag}
+                    labels = [label.id for label in obj.labels if label.tag == tag]
+                    if labels:
+                        return labels[0]
+                    return None
                 else:
-                    labels = {label.id: label.tag for label in obj.labels}
-                return labels
+                    return {label.id: label.tag for label in obj.labels}
         return {}
 
     def get_label_info(self, label_id: str) -> Dict[str, Any]:

